NXP LPSPI: Fix native CS on non-DMA driver

decsny opened the pull request:
Have to rework this driver completely to fix the CS because the SDK handle is based on the idea of a transfer being a single buffer, so we can't use the SDK transfer function

depends on:

- https://github.com/zephyrproject-rtos/zephyr/pull/82650
- https://github.com/zephyrproject-rtos/zephyr/pull/82806
- https://github.com/zephyrproject-rtos/zephyr/pull/82750
- https://github.com/zephyrproject-rtos/zephyr/pull/84406

Fixes #16544
Fixes #63280

Commit:
dts: nxp,lpspi: Fifo property and fix delay desc

Add properties for describing RX and TX fifo sizes.

Also reformat some descriptions and fix the description of the
transfer-delay property which was incorrect. Since zephyr spi bufs are
not continuous, every possible Zephyr LPSPI driver must use
continuous transfer mode, for which the meaning of this delay has
nothing to do with the chip select.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
spi_nxp_lpspi: Rewrite driver, fix native chip sel

To fix the native hardware chip select, we need to rewrite this driver
to not use the MCUX SDK handle abstraction, which does not fit the
zephyr use case.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

ofirshe commented:
Hi @decsny,

Thank you very much for your contribution on this front! I hope this issue will be resolved soon. I wanted to kindly ask if you’ve tested transferring and receiving with all possible combinations:
- When tx_len == rx_len
- When tx_len != rx_len
- When either tx_len or rx_len is zero.
As observed [here](https://github.com/zephyrproject-rtos/zephyr/discussions/81264#discussioncomment-11310063), the problem seems to occur when the sizes are equal in native GPIO CS mode, but it is expressed differently when the sizes are not equal.

Thanks again!

decsny commented:
> Hi @decsny,
> 
> Thank you very much for your contribution on this front! I hope this issue will be resolved soon. I wanted to kindly ask if you’ve tested transferring and receiving with all possible combinations:
> 
>     * When tx_len == rx_len
> 
>     * When tx_len != rx_len
> 
>     * When either tx_len or rx_len is zero.
>       As observed [here](https://github.com/zephyrproject-rtos/zephyr/discussions/81264#discussioncomment-11310063), the problem seems to occur when the sizes are equal in native GPIO CS mode, but it is expressed differently when the sizes are not equal.
> 
> 
> Thanks again!

Those cases are covered by the zephyr test suite. If you have more cases you want tested then please add them to the zephyr tests, would be useful

decsny commented:
The thing not covered by the zephyr test suite is the monitoring the chip select behavior, we might want to think about how to approach that

Commit:
drivers: spi_context: Functions for remaining len

Add functions to get remaining length in the transfer. Refactor a bit to
avoid duplicate code for the for loop that is the same as in the total
length function.

The difference between the total length and left length function is that
the current buffer total length is counted in the former and the current
buffer remaining length is counted in the latter.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

bperseghetti commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi.c:
```diff
@@ -0,0 +1,378 @@
+	spi_context_complete(ctx, dev, 0);
+}
+
+static inline uint8_t rx_fifo_cur_len(LPSPI_Type *base) {
```
```suggestion
static inline uint8_t rx_fifo_cur_len(LPSPI_Type *base)
{
```

bperseghetti reviewed (commented):
I think this is the only thing keeping it from passing the CI.

decsny commented:
rebased to fix conflicts

mmahadevan108 commented:
@wearyzen, can you help look at the IRQ additions.

mmahadevan108 commented:
@tbursztyka @teburd , can you help look at the changes to the SPI code.

wearyzen commented on drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi.c:
```diff
@@ -0,0 +1,362 @@
+
+	if (!spi_context_tx_on(ctx) && lpspi_is_rx_done(dev)) {
+		spi_context_complete(ctx, dev, 0);
+		arch_irq_clear_pending(config->irqn);
```
For consistency I think you should use irq_* macros as we have [here](https://github.com/zephyrproject-rtos/zephyr/blob/b80a058e9043087f2a114310df48eab4638598b6/include/zephyr/irq.h#L290) and use that here instead of calling arch_irq_*

wearyzen replied:
Sorry for not making that clear, what I meant if we are adding an arch_irq_* () apis then we should also add a new irq_clear_pending() macro in generic layer to be consistent with how arch_* apis are used.
Also, I see some other code in drivers using [NVIC_ClearPendingIRQ ](https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20NVIC_ClearPendingIRQ&type=code) and [NVIC_SetPendingIRQ ](https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20NVIC_SetPendingIRQ&type=code) so this is not the only use case.
You can chose to use the NVIC apis directly as others have used but if we add arch_irq_* apis then just defining these 3  seems incomplete. Ideally you would also need to define the arch_irq_* () to z_soc_irq_* () [here](https://github.com/zephyrproject-rtos/zephyr/pull/82877/files#diff-43e44cef4ccc0de92602731134c3013993fdae45d64a367dd80891185479f585R68) so things don't break when there is a custom interrupt controller defined with CONFIG_ARM_CUSTOM_INTERRUPT_CONTROLLER.

decsny replied:
okay, in this case, I will just directly use those CMSIS functions without any zephyr wrapper, because I don't want to address the arch irq code really in this PR, thanks

decsny commented:
updated to remove the commit touching the cortex m irq file, and use cmsis function directly instead, and rebased

jeppenodgaard commented:
Looks like this PR also fixes https://github.com/zephyrproject-rtos/zephyr/issues/63280.

teburd reviewed (approved):
Looks fine to me. I wouldn't bother implementing the transcieve_async API though as its not actually async (can block at any time, not usable from any context). To support rtio natively with this driver doesn't seem that far off, it requires using a different context with different functions to get the next read/write/transcieve and inform when the task is complete slightly different.
